google-datacatalog-sisense-connector/src/google/datacatalog_connectors/sisense/addons/elasticube_dependency_finder.py

```diff
@@ -0,0 +1,390 @@
+                entry_name)
+
+        return entries_dict
+
+    def __list_tags(self, entry_names: List[str]) -> Dict[str, List[Tag]]:
+        """Get Data Catalog Tags for a given list of entry names.
+
+        Returns:
+            dict: Keys are the entry names and values are lists that contains
+            all their related tags.
+        """
+        tags_dict = {}
+        for entry_name in entry_names or []:
+            tags_dict[entry_name] = self.__datacatalog_facade.list_tags(
+                entry_name)
+
+        return tags_dict
+
+    @classmethod
+    def __filter_relevant_tags(cls, entry: Entry, tags: List[Tag], table: str,
+                               column: str) -> List[Tag]:
+        """Filter Tags for ElastiCube dependencies finding. "Relevant", in this
+        context, means Tags that have ElastiCube-related information such as
+        data source, table, or column names.
+
+        Returns:
+            list: The filtered Tags.
+        """
+        filtered_tags = []
+
+        asset_metadata_tag = cls.filter_asset_metadata_tag(tags)
+        if asset_metadata_tag:
+            filtered_tags.append(asset_metadata_tag)
+        table_column_matching_tags = cls.__filter_table_column_matching_tags(
+            table, column, tags)
+        filtered_tags.extend(
+            cls.__sort_tags_by_schema(entry.schema.columns, '',
+                                      table_column_matching_tags))
+
+        return filtered_tags
+
+    @classmethod
+    def filter_asset_metadata_tag(cls, tags: List[Tag]) -> Optional[Tag]:
+        """Filter Dashboard and Widget metadata enrichment Tags created from
+        Templates identified by the `TAG_TEMPLATE_ID_DASHBOARD` and
+        `TAG_TEMPLATE_ID_WIDGET` constants.
+
+        Returns:
+            list: The filtered Tags.
+        """
+        template_ids = [
+            constants.TAG_TEMPLATE_ID_DASHBOARD,
+            constants.TAG_TEMPLATE_ID_WIDGET
+        ]
+
+        for tag in tags or []:
+            template_id = re.match(pattern=cls.__TAG_TEMPLATE_NAME_PATTERN,
+                                   string=tag.template).group('id')
+
+            if template_id in template_ids:
+                return tag
+
+    @classmethod
+    def __filter_table_column_matching_tags(cls, table: str, column: str,
+                                            tags: List[Tag]) -> List[Tag]:
+        """Filter JAQL-related Tags which `table` and/or `column` fields match
+        the provided values.
+
+        If both `table` and `column` args are provided, a given Tag's
+        corresponding fields must match both values in order to added to the
+        resulting list.
+
+        Returns:
+            list: The filtered Tags.
+        """
+        jaql_tags = cls.filter_jaql_tags(tags)
+        table_matching_tags = cls.__filter_table_matching_tags(
+            table, jaql_tags)
+        column_matching_tags = cls.__filter_column_matching_tags(
```

Comment:
The method logic, docstring and the method signature don't match:

1 -  Signature  
`__filter_table_column_matching_tags(cls, table: str, column: str...`  
table and column are required method args.

2 - Method logic  
```
        column_matching_tags = cls.__filter_column_matching_tags(
            column, table_matching_tags if table and column else tags)
```
Here table and column can be optional, so I'd suggest to either:
- Change method signature and use Optional[str], indicating that method args can be optional
- Change method logic to assume the args are always required

Reply:
There are other scenarios, I guess Optional is not really being used, so this could be addressed later in a separate PR. Since typing is being used in the codebase, would be nice to also explicitly type what's Optional and what is not...

This one caught my attention, because changes the method logic, it's not just a simple check like the other methods.

Reply:
The `table` and `column` args are actually optional, so I did small changes In order to declare them as named args with `None` as their default value wherever it makes sense. I decided on this approach because `Optional` is a type hint and ignoring it does not incur programming errors AFAIK (please correct me if I'm wrong).

Reply:
That is fine by me, using the default value makes it clear that the parameters are optional. Before my point was that it was confusing looking at the method args I'd think the args were required.

About type hints not incurring programming errors, you are right. But you can use the default args together with the type hints.

And you are already using those: `from typing import Dict, List, Optional, Tuple, Union`

If you are applying type hints, I'd prefer to use them all the way, we can add `mypy` to the build, and it would catch those typing errors.

Even `Optional` typing will be caught by it, let's say you mark an arg as Optional, and you don't handle the None value in the method, `mypy` will break the build and your IDE will also warn you. I've been using Optional heavily and it's helpful.

So to sum up, I'm ok if you prefer not to use it for now, but I'd suggest to use the default args as you are doing and also add Optional to them.

Reply:
Thanks for clarifying, @mesmacosta. I've added `Optional[str]` to the function args where it applies in the `ElastiCubeDependencyFinder` class. I intend to let the new code be as ready as possible for `mypy` 😃.

In regards to adding `mypy` checks to the CI, I'd rather tackle this in a separate MR because it could result in more changes to the whole code.

Reply:
yup! makes sense, will merge the MR then, and `mypy` can be handled in a separate MR.